lib/internal/modules/esm/default_resolve.js

```diff
@@ -51,17 +50,41 @@ if (experimentalWasmModules)
 if (experimentalJsonModules)
   extensionFormatMap['.json'] = legacyExtensionFormatMap['.json'] = 'json';
 
+function getFormat({ url, isMain }) {
+  if (NativeModule.canBeRequiredByUsers(url)) {
+    return { format: 'builtin' };
+  }
+  const parsed = new URL(url);
+  if (parsed.protocol === 'data:') {
+    const [ , mime ] = /^([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/.exec(parsed.pathname) || [ null, null, null ];
+    const format = ({
+      '__proto__': null,
+      'text/javascript': 'module',
+      'application/json': experimentalJsonModules ? 'json' : null,
+      'application/wasm': experimentalWasmModules ? 'wasm' : null
+    })[mime] || null;
+    return { format };
+  } else if (parsed.protocol === 'file:') {
+    const ext = extname(parsed.pathname);
+    let format = extensionFormatMap[ext];
+    if (ext === '.js' || (!format && isMain))
+      format = getPackageType(url.href) === TYPE_MODULE ? 'module' : 'commonjs';
+    if (!format && experimentalSpeciferResolution === 'node') {
+      process.emitWarning(
+        'The Node.js specifier resolution in ESM is experimental.',
+        'ExperimentalWarning');
+      format = legacyExtensionFormatMap[ext];
+    }
+    return { format: format || null };
+  }
+}
+
 function resolve(specifier, parentURL) {
+  const isMain = parentURL === undefined;
   try {
     const parsed = new URL(specifier);
     if (parsed.protocol === 'data:') {
-      const [ , mime ] = /^([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/.exec(parsed.pathname) || [ null, null, null ];
-      const format = ({
-        '__proto__': null,
-        'text/javascript': 'module',
-        'application/json': 'json',
-        'application/wasm': experimentalWasmModules ? 'wasm' : null
-      })[mime] || null;
+      const { format } = getFormat({ url: specifier, isMain });
```

Comment:
`getFormat` can be called twice - here and through `defaultGetFormat` of the custom `getFormat` hook.  Not sure this is a finding or just an observation, other than wasted cycles it doesn't look like calling the built-in `getFormat` twice is currently harmful.

Reply:
this is known and due to backwards compat, this PR is meant to be additive before we do any big breakage on loaders (which is where cleanup can occur all at once)